File: Chapter_2/MERGE_SORT.py
# Merge Sort implementaion without using sentinels
def Merge(A,p,q,r):
	n1=q-p+1
	n2=r-q
	L=[];R=[]
	for i in range(n1):
		L.append(A[p+i])
	for j in range(1,n2+1):
		R.append(A[q+j])
	for k in range(p,r+1):
		if(len(L)>0 and len(R)>0):
			if(L[0]<=R[0]):
				A[k]=L[0]
				del L[0]
			else:
				A[k]=R[0]
				del R[0]
		else:
			if(len(L)==0):
				A[k:r+1]=R
				break
			else:
				A[k:r+1]=L
				break
def Merge_Sort(A,p,r):
	if(p<r):
		q=(p+r)/2
		Merge_Sort(A,p,q)
		Merge_Sort(A,q+1,r)
		Merge(A,p,q,r)
if __name__ == '__main__':
	A=[9,8,7,6,5,4,3,2,1]
	print A
	Merge_Sort(A,0,len(A)-1)
	print A

File: Chapter_2/MergeSort.c
#include <stdio.h>
void Merge(int A[],int p,int q,int r){
	int n1 = q-p+1;
	int n2 = r-q;
	int L[n1],R[n2];
	for(int i=0;i<n1;i++) L[i] = A[p+i];
	for(int j=0;j<n2;j++) R[j] = A[q+j+1];
	int li=0,ri=0;
	for(int k=p;k<r+1;k++){
		if(li < n1 && ri < n2){
			if(L[li] <= R[ri]) A[k] = L[li++];
			else A[k] = R[ri++];
		}
		else{
			if(li == n1) A[k] = R[ri++];
			else A[k] = L[li++];
		}
	}
}
void mergeSort(int A[],int p,int r){
	if (p<r){
		int q = (p+r)/2;
		mergeSort(A,p,q);
		mergeSort(A,q+1,r);
		Merge(A,p,q,r);
	}
}
int main(){
	int A[9]={5,4,3,2,1,6,8,7,9};
	for(int i=0;i<9;i++) printf("%d ",A[i]);printf("\n");
	mergeSort(A,0,8);
	for(int i=0;i<9;i++) printf("%d ",A[i]);printf("\n");
	}